### PR TITLE
Zehua Week10 hw

### PR DESCRIPTION
**Pull request by zehua**

### CODE REVIEW

File: cis194/week10/zehua/src/AParser.hs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    where
+      pc s = runParser pa s >>= fb
+      fb (fa, sa) = runParser (fmap fa pb) sa
+  -}
+  {- using Parser Functor and Maybe Monad with do notation, probably the simplest
+  pa <*> pb = Parser pc
+    where
+      pc s = do
+        (fa, sa) <- runParser pa s
+        runParser (fmap fa pb) sa
```
Comment (zehua) [inline cis194/week10/zehua/src/AParser.hs]: is there a way to eliminate `s`?

File: cis194/week10/zehua/src/AParser.hs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  -}
+  {- using Parser Functor and manual Maybe mapping / flattening
+  pa <*> pb = Parser pc
+    where
+      pc s = oa
+        where
+          ra = runParser pa s
+          oa = case ra of
+                 Nothing -> Nothing
+                 Just (fa, sa) -> runParser (fmap fa pb) sa
```
Comment (zehua) [inline cis194/week10/zehua/src/AParser.hs]: not sure if it's common to do it in this type of nested `where` structure.

File: cis194/week10/zehua/src/AParser.hs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+-- Your code goes below here
+------------------------------------------------------------
+
+-- ex1
+first :: (a -> b) -> (a, c) -> (b, c)
+first f (a, c) = (f a, c)
+
+instance Functor Parser where
+  -- fmap :: (a -> b) -> Parser a -> Parser b
+  -- fmap :: (a -> b) -> (String -> Maybe (a, String)) -> (String -> Maybe (b, String))
```
Comment (zehua) [inline cis194/week10/zehua/src/AParser.hs]: this is technically not correct. just expanding it out to help understand the type

File: cis194/week10/zehua/src/AParser.hs (hunk 1 above)
Comment (shineli1984) [reply cis194/week10/zehua/src/AParser.hs]: There always is a way to eliminate all the argument. But I don’t think it’ll be pretty here

File: cis194/week10/zehua/src/AParser.hs (hunk 2 above)
Comment (shineli1984) [reply cis194/week10/zehua/src/AParser.hs]: Nested where should be fine. But in this case we can inline oa

**Pull request merged**